TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/Drivebase.java

```diff
@@ -0,0 +1,275 @@
+import com.qualcomm.robotcore.hardware.HardwareMap;
+import com.qualcomm.robotcore.hardware.IMU;
+
+import org.firstinspires.ftc.robotcore.external.Supplier;
+import org.firstinspires.ftc.robotcore.external.Telemetry;
+import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
+import org.firstinspires.ftc.teamcode.utility.Api;
+
+/**
+ * Class that contains the entire drivebase for the robot. Contains methods to move the robot both in auto and in teleop.
+ */
+public class Drivebase {
+    private final DcMotor frdrive;
+
+    private final DcMotor fldrive;
+    private final DcMotor brdrive;
+    private final DcMotor bldrive;
+    private final IMU imu;
+    private final Supplier<Boolean> opModeIsActive;
+
+
+    /**
+     * Uses `hardwareMap` to initialize the motors and imu.
+     *
+     * @param hardwareMap    The hardware map to use for initialization.
+     * @param opModeIsActive Supplier that returns false if the program needs to stop.
+     */
+    public @Api Drivebase(HardwareMap hardwareMap, Supplier<Boolean> opModeIsActive) {
+        frdrive = hardwareMap.get(DcMotor.class, "FRDrive");
+        fldrive = hardwareMap.get(DcMotor.class, "FLDrive");
+        brdrive = hardwareMap.get(DcMotor.class, "BRDrive");
+        bldrive = hardwareMap.get(DcMotor.class, "BLDrive");
+        imu = hardwareMap.get(IMU.class, "IMU");
+        this.opModeIsActive = opModeIsActive;
+
+        fldrive.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+        frdrive.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+        bldrive.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+        brdrive.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+
+        fldrive.setDirection(DcMotorSimple.Direction.REVERSE);
+        frdrive.setDirection(DcMotorSimple.Direction.FORWARD);
+        bldrive.setDirection(DcMotorSimple.Direction.REVERSE);
+        brdrive.setDirection(DcMotorSimple.Direction.FORWARD);
+
+        setMotorModes(DcMotor.RunMode.RUN_USING_ENCODER);
+    }
+
+    private void setMotorModes(DcMotor.RunMode mode) {
+        fldrive.setMode(mode);
+        frdrive.setMode(mode);
+        bldrive.setMode(mode);
+        brdrive.setMode(mode);
+    }
+
+    /**
+     * Adds the position of all the motors and the current heading to telemetry.
+     * Note that this is already called for you in auto, so you should only be calling this in teleop.
+     *
+     * @param telemetry The telemetry board to add the data to.
+     */
+    public @Api void addTelemetry(Telemetry telemetry) {
+        telemetry.addData("FLDrive", fldrive.getCurrentPosition());
+        telemetry.addData("FRDrive", frdrive.getCurrentPosition());
+        telemetry.addData("BLDrive", bldrive.getCurrentPosition());
+        telemetry.addData("BRDrive", brdrive.getCurrentPosition());
+        telemetry.addData("Heading", getHeading());
+        telemetry.update();
+    }
+
+    /**
+     * For use wherever.
```

Comment:
By having this function public, that's innately stating that this is for use wherever, you don't need to indicate it in your code. Following the principle of least privilege, you should only ever have access to what functions you **are** allowed to use in other files.

Reply:
I don't know how you would restrict this only to TeleOp without making things unnecessarily complicated. In our last project we had seperate TeleOpDrivebase and AutoDrivebase, made things a lot harder.